src/xmodem.cpp

```diff
@@ -0,0 +1,203 @@
+ * based on XMODEM implementation by Georges Menie (www.menie.org)
+ ***********************************************************************************************************************
+ * Copyright 2001-2019 Georges Menie (www.menie.org)
+ * All rights reserved.
+ * 
+ * Adapted for protobuf encapsulation. this is not really Xmodem any more.
+ * 
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the University of California, Berkeley nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE REGENTS AND CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **********************************************************************************************************************/
+
+#include "xmodem.h"
+
+XModemAdapter xModem;
+
+XModemAdapter::XModemAdapter()
+{
+}
+
+unsigned short XModemAdapter::crc16_ccitt(const pb_byte_t *buffer, int length)
+{
+    unsigned short crc16 = 0;
+    while(length != 0) {
+        crc16  = (unsigned char)(crc16 >> 8) | (crc16 << 8);
+        crc16 ^= *buffer;
+        crc16 ^= (unsigned char)(crc16 & 0xff) >> 4;
+        crc16 ^= (crc16 << 8) << 4;
+        crc16 ^= ((crc16 & 0xff) << 4) << 1;
+        buffer++;
+        length--;
+    }
+
+    return crc16;
+}
+
+int XModemAdapter::check(const pb_byte_t *buf, int sz, unsigned short tcrc)
+{
+    unsigned short crc = crc16_ccitt(buf, sz);
+    if (crc == tcrc)
+        return 1;
+    else
+        return 0;
+}
+
+void XModemAdapter::sendControl(XModem_Control c) {
+    memset(xmodemStore, 0, XModem_size);
+    xmodemStore->control = c;
+}
```

Comment:
Does anybody have a better idea than just copying stuff into xmodemStore and waiting for the PhoneAPI to find it? ToPhone queue seems to work the same way...